runner/providers/v0.1.1/external.go

```diff
@@ -0,0 +1,394 @@
+package v011
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"os/exec"
+
+	"github.com/pkg/errors"
+
+	garmErrors "github.com/cloudbase/garm-provider-common/errors"
+	commonExecution "github.com/cloudbase/garm-provider-common/execution/common"
+	commonParams "github.com/cloudbase/garm-provider-common/params"
+	garmExec "github.com/cloudbase/garm-provider-common/util/exec"
+	"github.com/cloudbase/garm/config"
+	"github.com/cloudbase/garm/metrics"
+	"github.com/cloudbase/garm/params"
+	"github.com/cloudbase/garm/runner/common"
+	"github.com/cloudbase/garm/runner/providers/util"
+)
+
+var _ common.Provider = (*external)(nil)
+
+func NewProvider(ctx context.Context, cfg *config.Provider, controllerID string) (common.Provider, error) {
+	if cfg.ProviderType != params.ExternalProvider {
+		return nil, garmErrors.NewBadRequestError("invalid provider config")
+	}
+
+	execPath, err := cfg.External.ExecutablePath()
+	if err != nil {
+		return nil, errors.Wrap(err, "fetching executable path")
+	}
+
+	// Set GARM_INTERFACE_VERSION to the version of the interface that the external
+	// provider implements. This is used to ensure compatibility between the external
+	// provider and garm
+	envVars := cfg.External.GetEnvironmentVariables()
+	envVars = append(envVars, fmt.Sprintf("GARM_INTERFACE_VERSION=%s", cfg.External.InterfaceVersion))
+
+	return &external{
+		ctx:                  ctx,
+		controllerID:         controllerID,
+		cfg:                  cfg,
+		execPath:             execPath,
+		environmentVariables: envVars,
+	}, nil
+}
+
+type external struct {
+	ctx                  context.Context
+	cfg                  *config.Provider
+	controllerID         string
+	execPath             string
+	environmentVariables []string
+}
+
+func (e *external) validateResult(inst commonParams.ProviderInstance) error {
+	if inst.ProviderID == "" {
+		return garmErrors.NewProviderError("missing provider ID")
+	}
+
+	if inst.Name == "" {
+		return garmErrors.NewProviderError("missing instance name")
+	}
+
+	if !util.IsValidProviderStatus(inst.Status) {
+		return garmErrors.NewProviderError("invalid status returned (%s)", inst.Status)
+	}
+
+	return nil
+}
+
+// CreateInstance creates a new compute instance in the provider.
+func (e *external) CreateInstance(ctx context.Context, bootstrapParams commonParams.BootstrapInstance, _ common.CreateInstanceParams) (commonParams.ProviderInstance, error) {
+	asEnv := []string{
+		fmt.Sprintf("GARM_COMMAND=%s", commonExecution.CreateInstanceCommand),
+		fmt.Sprintf("GARM_CONTROLLER_ID=%s", e.controllerID),
+		fmt.Sprintf("GARM_POOL_ID=%s", bootstrapParams.PoolID),
```

Comment:
to be consistent with the rest of the functions, I think we can have the `GARM_POOL_EXTRASPECS` variable set here as well. I know the `CreateInstance` command will look for extra specs in the bootstrap instance stdin body, but we should be consistent. The fewer differences in terms of env variables, the better.